File: Memory-OS-master/packages/sdk-java/src/main/java/com/memoryos/MemoryOS.java
package com.memoryos;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

/**
 * Official MEMORY OS Java SDK (v0.3). Model-independent; callers supply embeddings.
 */
public class MemoryOS {
    private final String baseUrl;
    private final String apiKey;
    private final HttpClient http = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(15)).build();
    private final ObjectMapper mapper = new ObjectMapper();

    public MemoryOS(String apiKey) {
        this(apiKey, System.getenv().getOrDefault("MEMORY_OS_API_URL", "http://localhost:8000"));
    }

    public MemoryOS(String apiKey, String baseUrl) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl;
    }

    public static class ApiException extends RuntimeException {
        public final String code;
        public final String requestId;
        public ApiException(String code, String message, String requestId) {
            super("[" + code + "] " + message + " (request_id=" + requestId + ")");
            this.code = code;
            this.requestId = requestId;
        }
    }

    public JsonNode createMemory(Map<String, Object> input) { return post("/v1/memory", input); }
    public JsonNode search(Map<String, Object> input) { return post("/v1/memory/search", input); }
    public JsonNode extract(Map<String, Object> input) { return post("/v1/memory/extract", input); }
    public JsonNode buildContext(Map<String, Object> input) { return post("/v1/context", input); }
    public JsonNode timeline(String memoryId) { return get("/v1/memory/" + memoryId + "/timeline"); }
    public JsonNode provenance(String memoryId) { return get("/v1/memory/" + memoryId + "/provenance"); }

    public JsonNode createSession(Map<String, Object> input) { return post("/v1/sessions", input); }
    public JsonNode listSessions(int limit, int offset) {
        return get("/v1/sessions?limit=" + limit + "&offset=" + offset);
    }
    public JsonNode sessionEvents(String sessionId) { return get("/v1/sessions/" + sessionId + "/events"); }
    public JsonNode appendSessionEvent(String sessionId, Map<String, Object> input) {
        return post("/v1/sessions/" + sessionId + "/events", input);
    }

    public JsonNode reflection(int staleDays) {
        return post("/v1/operations/reflection?stale_days=" + staleDays, Map.of());
    }
    public JsonNode reflectionExecute(Map<String, Object> input) {
        return post("/v1/operations/reflection/execute", input);
    }

    private JsonNode get(String path) { return request("GET", path, null); }
    private JsonNode post(String path, Map<String, Object> body) { return request("POST", path, body); }

    private JsonNode request(String method, String path, Map<String, Object> body) {
        try {
            HttpRequest.Builder b = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .timeout(Duration.ofSeconds(15))
                    .header("Content-Type", "application/json");
            if (apiKey != null && !apiKey.isEmpty()) {
                b.header("Authorization", "Bearer " + apiKey);
            }
            if ("POST".equals(method)) {
                String json = mapper.writeValueAsString(body != null ? body : Map.of());
                b.POST(HttpRequest.BodyPublishers.ofString(json));
            } else {
                b.GET();
            }
            HttpResponse<String> resp = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
            JsonNode node = mapper.readTree(resp.body());
            if (resp.statusCode() >= 400) {
                JsonNode err = node.get("error");
                throw new ApiException(
                        err.path("code").asText("UNKNOWN"),
                        err.path("message").asText("error"),
                        err.path("request_id").asText(null));
            }
            return node;
        } catch (ApiException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: Memory-OS-master/packages/sdk-kotlin/src/main/kotlin/com/memoryos/MemoryOS.kt
package com.memoryos

import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/** Official MEMORY OS Kotlin SDK (v0.3). Model-independent; callers supply embeddings. */
class MemoryOSException(val code: String, message: String, val requestId: String?) :
    RuntimeException("[$code] $message (request_id=$requestId)")

class MemoryOS(
    private val apiKey: String? = null,
    private val baseUrl: String = System.getenv("MEMORY_OS_API_URL") ?: "http://localhost:8000",
) {
    private val http = OkHttpClient.Builder()
        .callTimeout(15, TimeUnit.SECONDS).build()
    private val jsonMedia = "application/json".toMediaType()

    fun createMemory(body: JsonObject): JsonElement = post("/v1/memory", body)
    fun search(body: JsonObject): JsonElement = post("/v1/memory/search", body)
    fun extract(body: JsonObject): JsonElement = post("/v1/memory/extract", body)
    fun buildContext(body: JsonObject): JsonElement = post("/v1/context", body)
    fun timeline(memoryId: String): JsonElement = get("/v1/memory/$memoryId/timeline")
    fun provenance(memoryId: String): JsonElement = get("/v1/memory/$memoryId/provenance")

    fun createSession(body: JsonObject = buildJsonObject {}): JsonElement = post("/v1/sessions", body)
    fun listSessions(limit: Int = 25, offset: Int = 0): JsonElement =
        get("/v1/sessions?limit=$limit&offset=$offset")
    fun sessionEvents(sessionId: String): JsonElement = get("/v1/sessions/$sessionId/events")
    fun appendSessionEvent(sessionId: String, body: JsonObject): JsonElement =
        post("/v1/sessions/$sessionId/events", body)

    fun reflection(staleDays: Int = 90): JsonElement =
        post("/v1/operations/reflection?stale_days=$staleDays", buildJsonObject {})
    fun reflectionExecute(body: JsonObject): JsonElement =
        post("/v1/operations/reflection/execute", body)

    private fun get(path: String): JsonElement = request("GET", path, null)
    private fun post(path: String, body: JsonObject): JsonElement = request("POST", path, body)

    private fun request(method: String, path: String, body: JsonObject?): JsonElement {
        val reqBuilder = Request.Builder().url("$baseUrl$path")
            .header("Content-Type", "application/json")
        apiKey?.let { reqBuilder.header("Authorization", "Bearer $it") }
        if (method == "POST") {
            reqBuilder.post((body ?: buildJsonObject {}).toString().toRequestBody(jsonMedia))
        } else {
            reqBuilder.get()
        }

        http.newCall(reqBuilder.build()).execute().use { resp ->
            val text = resp.body?.string() ?: "{}"
            val node = Json.parseToJsonElement(text)
            if (!resp.isSuccessful) {
                val err = node.jsonObject["error"]!!.jsonObject
                throw MemoryOSException(
                    err["code"]?.jsonPrimitive?.content ?: "UNKNOWN",
                    err["message"]?.jsonPrimitive?.content ?: "error",
                    err["request_id"]?.jsonPrimitive?.contentOrNull,
                )
            }
            return node
        }
    }
}
